packages/parser/src/index.tsx

```diff
@@ -0,0 +1,43 @@
+import { JSXElement, createMemo } from "solid-js";
+
+export function createParser(id?: string) {
+  const $TOKEN = Symbol(id || "solid-parser");
+
+  function tokenize<
+    TToken extends Meta & {
+      props: unknown;
+    }
+  >(meta: Omit<TToken, "props">): (props: TToken["props"]) => JSXElement {
+    return (props: any) => {
+      return Object.assign(
+        () => {
+          console.info("tokens can only be rendered inside a Parser");
+          return <></>;
+        },
+        {
+          [$TOKEN]: true,
+          ...meta,
+          props
+        }
+      );
+    };
+  }
+
+  function childrenTokens<T>(children: () => JSXElement | JSXElement[] | T | T[]) {
```

Comment:
Not sure about the generic here
I guess I can help with typings once it's finished
```suggestion
  function childrenTokens<T = {}>(children: () => JSX.Element): T[] {
```

Reply:
it is a bit awkward having to pass in T like that, especially because that also means u have to type the children like that as well from the component you want to do `childrenTokens`, but without it I can not get `childrenTokens` to return the proper type.

```ts
type Token = {id: 'token'}
const Parser = (props: { children: JSXElement | JSXElement[] }) => {
  const tokens = childrenTokens(() => props.children); 
  tokens[0].id  // type-error: token is of type unknown
}

const Parser = (props: { children: JSXElement | JSXElement[] | Token | Token[] }) => {
  const tokens = childrenTokens(() => props.children); 
  tokens[0].id  // no type-error
}
```

maybe a type-helper could help here?

```
type Children<TToken> = JSXElement | JSXElement[] | TToken | TToken[]
const Parser = (props: { children: Children<Token> }) => {
  const tokens = childrenTokens(() => props.children); 
  tokens[0].id  // no type-error
}
```

Reply:
O you mean probably that you would type every `childrenToken`?

like
```
const Parser = (props: { children: JSXElement | JSXElement[] }) => {
  const tokens = childrenTokens<Token>(() => props.children); 
  tokens[0].id  // no type-error
}
```
that would work, is a bit verbose maybe.

ye, that casting of the token is a bit tricky actually mm...

Reply:
maybe instead of typing childrenTokens and hoping that the dev enters the correct generic, we could `createJSXParser` accept a generic of `CustomToken` which is a union of all the possible types, and that's what is being returned.

```ts
type Props = {brightness: number}
type GreenToken = {id: 'green', props: Props}
type RedToken = {id: 'red', props: Props}

type CustomToken  = GreenToken | RedToken

const {createToken, childrenTokens} = createParser<CustomToken>('color-parser')

const Green = createToken<Props, GreenToken>(props => ({props, id: 'green'})
const Red = createToken<Props, RedToken>(props => ({props, id: 'red'})

const Colors = (props: {children: JSXElement | JSXElement[]) => {
   const tokens = childrenTokens(() => props.children)
   if(token.id === 'red')
       token // RedToken
}

```

Reply:
yes! i think this is the solution. we should then also be possible to type the returntype of the callback in `childrenTokens(callback)` so that it fits the generic of `createJSXParser`.

Reply:
oo, that also means that even if you don't provide a Token-generic in `createToken`, it will still typecheck if what you return from the callback fits any of the accepted types.

```
type Props = {brightness: number}
type GreenToken = {id: 'green', props: Props}
type RedToken = {id: 'red', props: Props}

type CustomToken  = GreenToken | RedToken

const {createToken, childrenTokens} = createJSXParser<CustomToken>('color-parser')

const Yellow = createToken(props => ({props, id: 'yellow'}) 
  // type-error: Type 'yellow' is not assignable to type 'green' | 'red'
```

Reply:
[can we do it without extends though?](https://github.com/solidjs-community/solid-primitives/pull/276/commits/e35fac886b7242885769b1a70af0e1b9b145c322#r1061866399)